drivers/entropy/entropy_stm32.c

```diff
@@ -347,25 +347,55 @@ static int entropy_stm32_rng_get_entropy_isr(const struct device *dev,
 	return cnt;
 }
 
-static int entropy_stm32_rng_init(const struct device *dev)
+#if CONFIG_SOC_SERIES_STM32L4X
+static int entropy_stm32_rng_init_l4(void)
 {
-	struct entropy_stm32_rng_dev_data *dev_data;
-	const struct entropy_stm32_rng_dev_cfg *dev_cfg;
-	int res;
-
-	__ASSERT_NO_MSG(dev != NULL);
-
-	dev_data = DEV_DATA(dev);
-	dev_cfg = DEV_CFG(dev);
-
-	__ASSERT_NO_MSG(dev_data != NULL);
-	__ASSERT_NO_MSG(dev_cfg != NULL);
+	/*
+	 * PLL configuration table for all possible MSI input frequencies. Parameters
+	 * are chosen to obtain a 48 MHz clock for USB FS, SDMMC and RNG. For the
+	 * lower frequencies the highest possible output value is chosen that can be
+	 * used for RNG.
+	 */
+	static const struct {
+		uint8_t pllm;
+		uint8_t plln;
+		uint32_t pllq;
+	} pll_msi_cfg[] = {
+		/*PLLM               N   PLLMQ*/                  /*clkin   PLL48M2CLK*/
+		{ LL_RCC_PLLM_DIV_1, 86, LL_RCC_PLLSAI1Q_DIV_2 }, /*0.1 MHz 4.3 MHz   */
+		{ LL_RCC_PLLM_DIV_1, 86, LL_RCC_PLLSAI1Q_DIV_2 }, /*0.2 MHz 8.6 MHz   */
+		{ LL_RCC_PLLM_DIV_1, 86, LL_RCC_PLLSAI1Q_DIV_2 }, /*0.4 MHz 17.2 MHz  */
+		{ LL_RCC_PLLM_DIV_1, 86, LL_RCC_PLLSAI1Q_DIV_2 }, /*0.8 MHz 34.4 MHz  */
+		{ LL_RCC_PLLM_DIV_1, 48, LL_RCC_PLLSAI1Q_DIV_2 }, /*1 MHz   43   MHz  */
+		{ LL_RCC_PLLM_DIV_1, 24, LL_RCC_PLLSAI1Q_DIV_2 }, /*2 MHz   48   MHz  */
+		{ LL_RCC_PLLM_DIV_1, 24, LL_RCC_PLLSAI1Q_DIV_2 }, /*4 MHz   48   MHz  */
+		{ LL_RCC_PLLM_DIV_1, 24, LL_RCC_PLLSAI1Q_DIV_4 }, /*8 MHz   48   MHz  */
+		{ LL_RCC_PLLM_DIV_1, 24, LL_RCC_PLLSAI1Q_DIV_8 }, /*16 MHz  48   MHz  */
+		{ LL_RCC_PLLM_DIV_2, 32, LL_RCC_PLLSAI1Q_DIV_8 }, /*12 MHz  48   MHz  */
+		{ LL_RCC_PLLM_DIV_2, 24, LL_RCC_PLLSAI1Q_DIV_8 }, /*16 MHz  48   MHz  */
+		{ LL_RCC_PLLM_DIV_2, 16, LL_RCC_PLLSAI1Q_DIV_8 }, /*24 MHz  48   MHz  */
+	};
+	uint32_t msi_range;
+
+	if (LL_RCC_MSI_IsEnabledRangeSelect() == 0) {
+		msi_range = LL_RCC_MSI_GetRangeAfterStandby();
+		if (msi_range < LL_RCC_MSISRANGE_4 || msi_range > LL_RCC_MSISRANGE_7) {
+			return -EINVAL;
+		}
+		msi_range >>= RCC_CSR_MSISRANGE_Pos;
+	} else {
+		msi_range = LL_RCC_MSI_GetRange();
+		if (msi_range > LL_RCC_MSIRANGE_11) {
+			return -EINVAL;
+		}
+		msi_range >>= RCC_CR_MSIRANGE_Pos;
+	}
+	__ASSERT_NO_MSG(msi_range < ARRAY_SIZE(pll_msi_cfg));
```

Comment:
PLL should not be configured in this driver, this should be done in the dedicated driver (clock-control).
Also, for now for all PLLs we rely on user to provide the correct configuration using device tree, this should be the same for this one.

Reply:
The PLL configuration was already in this driver ("Configure PLLSA11 to enable 48M domain"). I just ha to fix it when using other MSI frequencies.
I'm not familiar with clock-control functions. Do you mean that the clock selection should be done in following clock_control_on() or something similar? Where is the PLLSAI1 clock selected in the device tree?
Should we solve this issue in this PR or does it make sense to move it out?

Reply:
> The PLL configuration was already in this driver ("Configure PLLSA11 to enable 48M domain"). I just ha to fix it when using other MSI frequencies.

Indeed, but this was just as a convenience. I'd like we put too much code for PLL support in drivers other than clocks.
This need to be centralized, for architecture coherency, ease of maintenance and footprint.

> I'm not familiar with clock-control functions. Do you mean that the clock selection should be done in following clock_control_on() or something similar? Where is the PLLSAI1 clock selected in the device tree?

We surely miss code for that now, but this should be added. Clock control description has been moved to dts recently as a preliminary step in that direction. I'd like to avoid having drivers going in the opposite direction from now on.

> Should we solve this issue in this PR or does it make sense to move it out?

This should probably be done in another PR

Reply:
I totally agree to have cleaner architecture, moving the code in proper drivers, have common code for different families, etc. We seem even to agree that the scope of this PR is fixing some non-working cases, limited to our experience with L462VE.